includes/class-wporg-gp-translation-events-stats-calculator.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+class WPORG_GP_Translation_Events_Stats_Row {
+	public int $created;
+	public int $reviewed;
+	public int $users;
+
+	public function __construct( $created, $reviewed, $users ) {
+		$this->created  = $created;
+		$this->reviewed = $reviewed;
+		$this->users    = $users;
+	}
+}
+
+class WPORG_GP_Translation_Events_Event_Stats {
+	/**
+	 * Associative array of rows, with the locale as key.
+	 *
+	 * @var WPORG_GP_Translation_Events_Stats_Row[]
+	 */
+	private array $rows = [];
+
+	private WPORG_GP_Translation_Events_Stats_Row $totals;
+
+	/**
+	 * @throws Exception
+	 */
+	public function add_row( string $locale, WPORG_GP_Translation_Events_Stats_Row $row ) {
+		if ( ! $locale ) {
+			throw new Exception( 'locale must not be empty' );
+		}
+		$this->rows[ $locale ] = $row;
+	}
+
+	public function set_totals( WPORG_GP_Translation_Events_Stats_Row $totals ) {
+		$this->totals = $totals;
+	}
+
+	/**
+	 * Get an associative array of rows, with the locale as key.
+	 *
+	 * @return WPORG_GP_Translation_Events_Stats_Row[]
+	 */
+	public function rows(): array {
+		return $this->rows;
+	}
+
+	public function totals(): WPORG_GP_Translation_Events_Stats_Row {
+		return $this->totals;
+	}
+}
+
+class WPORG_GP_Translation_Events_Stats_Calculator {
+	private string $ACTIONS_TABLE_NAME = WPORG_GP_Translation_Events_Translation_Listener::ACTIONS_TABLE_NAME;
+
+	/**
+	 * @throws Exception
+	 */
+	function for_event( WP_Post $event ): WPORG_GP_Translation_Events_Event_Stats {
+		$start = new DateTime( get_post_meta( $event->ID, '_event_start', true ), new DateTimeZone( 'UTC' ) );
+		$end   = new DateTime( get_post_meta( $event->ID, '_event_end', true ), new DateTimeZone( 'UTC' ) );
+		$stats = new WPORG_GP_Translation_Events_Event_Stats;
+		global $wpdb;
+
+		$query = $wpdb->prepare( "
+				select locale,
+					   sum(if(action = 'create', 1, 0)) AS created,
+					   sum(if(action in ('approve', 'reject', 'request_changes', 'mark_fuzzy'), 1, 0)) AS reviewed,
+					   count(distinct user_id) as users
+				from $this->ACTIONS_TABLE_NAME
+				where event_id = %d
+				  and happened_at between cast('%s' as datetime) and cast('%s' as datetime)
+				group by locale with rollup
```

Comment:
Because of this we should add an `event_id, locale` index to the table so that doesn't have to do filesorting.

Reply:
While you're adding the index, could you please also make the primary index `event_id, user_id, translation_id`? I think it's much more likely that we might later want to do a query of all translations a certain user did in an event than which users have an entry on a translation id (which at most would be two). MySQL can only use the index in sequence, so when querying for `event_id` and `user_id`, it can only use the `event_id` part of an `event_id, translation_id, user_id` index but `event_id` and `user_id` for a `event_id, user_id, translation_id` index.

Reply:
Ah, so that's where the filesorting is coming from, thanks! Agree on changing the order of columns in the index too, will fix both these issues.